src/server.ts

```diff
@@ -70,43 +72,195 @@ export async function buildServer (ctx: ViteBuildContext) {
-      if (event.code === 'BUNDLE_START') {
-        start = Date.now()
-      } else if (event.code === 'BUNDLE_END') {
-        await generateDevSSRManifest(ctx)
-        await onBuild()
-        consola.info(`Server rebuilt in ${Date.now() - start}ms`)
-      } else if (event.code === 'ERROR') {
-        consola.error(event.error)
+    return
+  }
+
+  // Start development server
+  const viteServer = await vite.createServer(serverConfig)
+  // Initialize plugins
+  await viteServer.pluginContainer.buildStart({})
+
+  const { code: esm } = await bundleRequest(viteServer, '/.nuxt/server.js')
+
+  // FIXME: vue-bundle-renderer does not support ESM
+  const cjs = `
+module.exports = async (ctx) => {
+  // const server = await import('${resolve(ctx.nuxt.options.buildDir, 'dist/server/server.mjs')}')
+  const server = require('jiti')()('${resolve(ctx.nuxt.options.buildDir, 'dist/server/server.mjs')}')
+  const result = await server.default().then(i => i.default(ctx))
+  return result
+}`
+
+  await writeFile(resolve(ctx.nuxt.options.buildDir, 'dist/server/server.mjs'), esm, 'utf-8')
+  await writeFile(resolve(ctx.nuxt.options.buildDir, 'dist/server/server.js'), cjs, 'utf-8')
+
+  await writeFile(resolve(ctx.nuxt.options.buildDir, 'dist/server/ssr-manifest.json'), JSON.stringify({}, null, 2), 'utf-8')
+
+  await generateDevSSRManifest(ctx)
+  await onBuild()
+  ctx.nuxt.hook('close', () => viteServer.close())
+}
+
+// ---- Vite Dev Bundler POC ----
+
+interface TransformChunk {
+  id: string,
+  code: string,
+  deps: string[],
+  parents: string[]
+}
+
+interface SSRTransformResult {
+  code: string,
+  map: object,
+  deps: string[]
+}
+
+async function transformRequest (viteServer: vite.ViteDevServer, id) {
+  // Virtual modules start with `\0`
+  if (id && id.startsWith('/@id/__x00__')) {
+    id = '\0' + id.slice('/@id/__x00__'.length)
+  }
+
+  // Externals
+  if (builtinModules.includes(id) || (id.includes('node_modules') && !id.endsWith('.esm.js'))) {
+    if (id.startsWith('/@fs')) {
+      id = id.substr(4)
+    } else if (id.startsWith('/')) {
+      id = '.' + id // TODO
+    }
+    return {
+      code: `() => import('${id}')`,
+      deps: []
+    }
+  }
+
+  // Transform
+  const res: SSRTransformResult = await viteServer.transformRequest(id, { ssr: true }).catch((err) => {
+    // eslint-disable-next-line no-console
+    console.warn(`[SSR] Error transforming ${id}: ${err}`)
+    // console.error(err)
+  }) as SSRTransformResult || { code: '', map: {}, deps: [] }
+
+  // Wrap into a vite module
+  const code = `async function () {
+const __vite_ssr_exports__ = {};
+const __vite_ssr_exportAll__ = __createViteSSRExportAll__(__vite_ssr_exports__)
+${res.code || '/* empty */'};
+return __vite_ssr_exports__;
+}`
+  return { code, deps: res.deps || [] }
+}
+
+function scanDynamicDeps (code: string) {
+  return Array.from(code.matchAll(/__vite_ssr_dynamic_import__\(['"](.+?)['"]\)/g)).map(m => m[1])
```

Comment:
@antfu Don't we have access to this via `deps`? If not maybe as an improvement vite can return `dynamicDeps` array to avoid searching string?

Reply:
Yes, we don't. The `deps` is used for Vite to know the HMR boundaries. I think it assumes dynamically imported modules as separate requests and does not include them intentionally. Meanwhile, `transformRequest` returns rollup compatible `TransformResult` which has only the deps field. If we want `dynamicDeps`, maybe we will need to extend the interface for it.

I will open an issue for Vite and discuss it with the team this evening. Meanwhile, giving this identifier are quite special already, I guess we can safely rely on this workaround for the current stage.

Reply:
PR is here: https://github.com/vitejs/vite/pull/4909

Reply:
...and it's shipped!